{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "76b86bb5",
   "metadata": {},
   "source": [
    "https://rosalind.info/problems/orf/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f6fb471",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MWLAYLSGSLKECRKTWPADFISPAINLALSIIFGAMGQFLASH\n",
      "M\n",
      "MGQFLASH\n",
      "MVEIAMSLPWVDNVLNDCQYFSGYVLPPKMLLSFAAGVFLLVPRRRDRMSSALDLNYVNPLKLRLKLRRRTLSHEPKRYA\n",
      "MSLPWVDNVLNDCQYFSGYVLPPKMLLSFAAGVFLLVPRRRDRMSSALDLNYVNPLKLRLKLRRRTLSHEPKRYA\n",
      "MCLTTVSISRGTFYRQKCYYRSQQACSYSFRGVVIE\n",
      "MLLSFAAGVFLLVPRRRDRMSSALDLNYVNPLKLRLKLRRRTLSHEPKRYA\n",
      "MSSALDLNYVNPLKLRLKLRRRTLSHEPKRYA\n",
      "MNLSGMHNLL\n",
      "MHNLL\n",
      "MTCLPT\n",
      "MWRQSRTPSEPLSTHRHPVQ\n",
      "MAVSIRDQLSELGRVAALSRKEQDRKPSLYWMSVGR\n",
      "MSVGR\n",
      "MSQTDS\n",
      "MPIRCNDLLANLAKLASKS\n",
      "MQ\n",
      "MHTA\n",
      "MRKRPPSQLETKFKRIDVIKVERAAHSITTPRNE\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "file = \"sample.txt\"\n",
    "\n",
    "genes = list(SeqIO.parse(file, \"fasta\"))\n",
    "dna_seq = \"\"\n",
    "rna_seq1 = \"\"\n",
    "rna_seq2 = \"\"\n",
    "\n",
    "for i in range(0, len(genes)):\n",
    "    dna_seq = str(genes[i].seq)\n",
    "\n",
    "for n in dna_seq:\n",
    "    rna_seq1 = dna_seq.replace(\"T\", \"U\") \n",
    "\n",
    "def rev_comp(dna):\n",
    "    complement = dict(zip(\"AUGC\", \"UACG\"))\n",
    "    new_dna = \"\"\n",
    "    for l in dna:\n",
    "        new_dna += complement[l]\n",
    "    return new_dna[::-1]\n",
    "\n",
    "rna_seq2 = rev_comp(rna_seq1)\n",
    "\n",
    "def find_orf(rna1, rna2):\n",
    "    rna_codon_table = {\n",
    "        'UUU': 'F',     'CUU': 'L',     'AUU': 'I',     'GUU': 'V',\n",
    "        'UUC': 'F',     'CUC': 'L',     'AUC': 'I',     'GUC': 'V',\n",
    "        'UUA': 'L',     'CUA': 'L',     'AUA': 'I',     'GUA': 'V',\n",
    "        'UUG': 'L',     'CUG': 'L',     'AUG': 'M',     'GUG': 'V',\n",
    "        'UCU': 'S',     'CCU': 'P',     'ACU': 'T',     'GCU': 'A',\n",
    "        'UCC': 'S',     'CCC': 'P',     'ACC': 'T',     'GCC': 'A',\n",
    "        'UCA': 'S',     'CCA': 'P',     'ACA': 'T',     'GCA': 'A',\n",
    "        'UCG': 'S',     'CCG': 'P',     'ACG': 'T',     'GCG': 'A',\n",
    "        'UAU': 'Y',     'CAU': 'H',     'AAU': 'N',     'GAU': 'D',\n",
    "        'UAC': 'Y',     'CAC': 'H',     'AAC': 'N',     'GAC': 'D',\n",
    "        'UAA': 'Stop',  'CAA': 'Q',     'AAA': 'K',     'GAA': 'E',\n",
    "        'UAG': 'Stop',  'CAG': 'Q',     'AAG': 'K',     'GAG': 'E',\n",
    "        'UGU': 'C',     'CGU': 'R',     'AGU': 'S',     'GGU': 'G',\n",
    "        'UGC': 'C',     'CGC': 'R',     'AGC': 'S',     'GGC': 'G',\n",
    "        'UGA': 'Stop',  'CGA': 'R',     'AGA': 'R',     'GGA': 'G',\n",
    "        'UGG': 'W',     'CGG': 'R',     'AGG': 'R',     'GGG': 'G'\n",
    "    }\n",
    "\n",
    "    strands = [rna1, rna2]\n",
    "    orf_seq = []\n",
    "    orf_list = []\n",
    "    \n",
    "    for strand in strands:\n",
    "        for i in range(0, len(strand)):\n",
    "            if strand[i:i+3] == \"AUG\":\n",
    "                new_strand = strand[i:-1]\n",
    "                for j in range(0, len(new_strand)-3, 3):\n",
    "                    if new_strand[j:j+3] in ('UAA', 'UAG', 'UGA'):\n",
    "                        scan_seq = new_strand[0:j+3]\n",
    "                        orf_seq.append(scan_seq)\n",
    "                        break\n",
    "\n",
    "    for i in range(0, len(orf_seq)):\n",
    "        aa = \"\"\n",
    "        scan_seq = orf_seq[i]\n",
    "        for j in range(0, len(scan_seq), 3):\n",
    "            codon = scan_seq[j:j+3]\n",
    "            tRNA_aa = rna_codon_table[codon]\n",
    "            if tRNA_aa != \"Stop\":\n",
    "                aa += tRNA_aa\n",
    "        if aa not in orf_list:\n",
    "            orf_list.append(aa)\n",
    "\n",
    "    return orf_list\n",
    "\n",
    "result = find_orf(rna_seq1, rna_seq2)\n",
    "\n",
    "for n in result:\n",
    "    print(n)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac9d2e8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
